internal/xds/rbac/rbac_engine.go

```diff
@@ -142,41 +142,65 @@ func newEngine(config *v3rbacpb.RBAC, policyName string) (*engine, error) {
 		policies[name] = matcher
 	}
 
-	auditOptions := config.GetAuditLoggingOptions()
-	auditLoggers := []audit.Logger{}
-	if auditOptions != nil {
-		for _, logger := range auditOptions.LoggerConfigs {
-			auditLoggerFactory := audit.GetLoggerBuilder(logger.AuditLogger.Name)
-			if auditLoggerFactory == nil {
-				if logger.IsOptional {
-					continue
-				}
-				return nil, fmt.Errorf("no builder registered for %v", logger.AuditLogger.Name)
-			}
-			auditLoggerConfig, err := parseCustomConfig(auditLoggerFactory, logger.AuditLogger.TypedConfig)
-			if err != nil {
-				return nil, fmt.Errorf("audit logger custom config could not be parsed: %v", err)
-			}
-			auditLogger := auditLoggerFactory.Build(auditLoggerConfig)
-			auditLoggers = append(auditLoggers, auditLogger)
-		}
+	auditLoggers, auditCondition, err := parseAuditOptions(config.GetAuditLoggingOptions())
+	if err != nil {
+		return nil, err
 	}
 	return &engine{
 		policies:       policies,
 		action:         a,
 		auditLoggers:   auditLoggers,
-		auditCondition: auditOptions.GetAuditCondition(),
+		auditCondition: auditCondition,
 		policyName:     policyName,
 	}, nil
 }
 
-func parseCustomConfig(factory audit.LoggerBuilder, pb *anypb.Any) (audit.LoggerConfig, error) {
-	st := new(structpb.Struct)
-	err := pb.UnmarshalTo(st)
+func parseAuditOptions(opts *v3rbacpb.RBAC_AuditLoggingOptions) ([]audit.Logger, v3rbacpb.RBAC_AuditLoggingOptions_AuditCondition, error) {
+	if opts == nil {
+		return nil, v3rbacpb.RBAC_AuditLoggingOptions_NONE, nil
+	}
+	var auditLoggers []audit.Logger
+	for _, logger := range opts.LoggerConfigs {
+		if logger.AuditLogger.TypedConfig == nil {
+			return nil, v3rbacpb.RBAC_AuditLoggingOptions_NONE, fmt.Errorf("AuditLogger TypedConfig cannot be nil")
+		}
+		customConfig, err := getCustomConfig(logger.AuditLogger.TypedConfig)
+		if err != nil {
+			return nil, v3rbacpb.RBAC_AuditLoggingOptions_NONE, err
+		}
+		if customConfig.GetTypeUrl() == "" {
+			return nil, v3rbacpb.RBAC_AuditLoggingOptions_NONE, fmt.Errorf("AuditLogger TypedConfig.TypeURL cannot be an empty string")
+		}
+
+		auditLoggerFactory := audit.GetLoggerBuilder(customConfig.GetTypeUrl())
```

Comment:
We shouldn't use the `type_url` directly to get the builder. Users will register the logger with the name they use in the authz policy. They are not supposed to know the `grpc.audit_logging/` prefix. Therefore, we need to strip the prefix and check if there is anything left after the last `/`.

This logic doesn't have to reside in this pkg. Something like https://github.com/easwars/grpc-go/blob/master/xds/internal/xdsclient/xdslbregistry/converter.go is better to be added and invoked here. cc @erm-g. As such, this `parseAuditOptions` doesn't have to prepare the config for the builder by itself.

Let me know if there are any questions.

Reply:
Discussed offline, PTAL at newest changes